SIGSEGV linuxfw.getChainFromTable

### What is the issue?

Tailscale [1.98.1 static binary](https://pkgs.tailscale.com/stable/tailscale_1.98.1_arm.tgz) on a [JetKVM](https://jetkvm.com/) crashes on start.

```
TPM: error opening: stat /dev/tpmrm0: no such file or directory
logtail started
Program starting: v1.98.1-t2f4c02ee6-g0cd07359c, Go 1.26.2: []string{"/userdata/tailscale/tailscaled", "-port", "41641"}
LogID: […]
dns: using dns.noopManager
cleanup: list tables: socket: protocol not supported
wgengine.NewUserspaceEngine(tun "tailscale0") ...
dns: using dns.noopManager
link state: interfaces.State{defaultRoute=eth0 ifs={eth0:[192.0.2.24/24 llu6]} v4=true v6=false}
magicsock: disco key = d:c7088155102f7c74
Creating WireGuard device...
Bringing WireGuard device up...
Bringing router up...
router: netfilter running in nftables mode, v6 = true
router: portUpdate(port=41641, network=udp6)
external route: up
Clearing router settings...
Starting network monitor...
Engine created.
router: portUpdate(port=41641, network=udp4)
monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:254 Protocol:0 Scope:0 Type:1 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5210 Table:254 Mark:16711680 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:253 Protocol:0 Scope:0 Type:1 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5230 Table:253 Mark:16711680 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:0 Protocol:0 Scope:0 Type:7 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5250 Table:0 Mark:16711680 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:52 Protocol:0 Scope:0 Type:1 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5270 Table:52 Mark:0 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
monitor: gateway and self IP changed: gw=192.0.2.1 self=192.0.2.24
pm: using backend prefs for "profile-9b9a": Prefs{ra=false dns=true want=true routes=[] statefulFiltering=false nf=on update=on Persist{o=, n=[EeY0P] u="host.tail.ts.net" ak=-}}
linkChange: in state NoState; PAC or proxyConfig changed; updating routes
got LocalBackend in 202ms
Start
ipnext: "conn25": skipping extension
offline auto-update: starting update checks
ipnext: active extensions: posture, clientupdate, relayserver, taildrop, conn25, portlist
load netmap from cache: netmap cache is not available
tka initialized at head […]
control: tkaHead: […]
Backend: logs: be:[…] fe:
control: client.Login(0)
control: doLogin(regen=false, hasUrl=false)
health(warnable=warming-up): error: Tailscale is starting. Please wait.
router: somebody (likely systemd-networkd) deleted ip rules; restoring Tailscale's
control: control server key from https://controlplane.tailscale.com: ts2021=[fSeS+], legacy=[nlFWp]
control: RegisterReq: onode= node=[EeY0P] fup=false nks=false
control: RegisterReq: got response; nodeKeyExpired=false, machineAuthorized=true; authURL=false
health(warnable=not-in-map-poll): ok
control: netmap: got new dial plan from control
tkaSyncIfNeeded: isEnabled=true, wantEnabled=true, nodeHead=[…], netmapHead=[…]
active login: host.tail.ts.net
netmap: suggested exit node: no preferred DERP, try again later
offline auto-update: stopping update checks
Switching ipn state NoState -> Starting (WantRunning=true, nm=true)
magicsock: SetPrivateKey called (init)
wgengine: Reconfig: configuring userspace WireGuard config (with 3 peers)
wgengine: Reconfig: configuring router
router: enabling connmark-based rp_filter workaround
router: warning: failed to add connmark rules (rp_filter workaround may not work): flush add connmark rules: socket: protocol not supported
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xa23208]

goroutine 160 [running]:
tailscale.com/util/linuxfw.getChainFromTable(0x3912ff0, 0x0, {0xdc7b94, 0x8})
	tailscale.com@v1.98.1/util/linuxfw/nftables_runner.go:419 +0x14
tailscale.com/util/linuxfw.(*nftablesRunner).AddExternalCGNATRules(0x3811120, {0xdc2440, 0x4}, {0x38135c0, 0xa})
	tailscale.com@v1.98.1/util/linuxfw/nftables_runner.go:1597 +0x40
tailscale.com/wgengine/router/osrouter.(*linuxRouter).setCGNATDropModeLocked(0x38b9980, {0xdc2440, 0x4})
	tailscale.com@v1.98.1/wgengine/router/osrouter/router_linux.go:566 +0x164
tailscale.com/wgengine/router/osrouter.(*linuxRouter).Set(0x38b9980, 0x6579a90)
	tailscale.com@v1.98.1/wgengine/router/osrouter/router_linux.go:541 +0x112c
tailscale.com/wgengine.(*userspaceEngine).Reconfig(0x3bfc008, 0x656f368, 0x6579a90, 0x6571230)
	tailscale.com@v1.98.1/wgengine/userspace.go:980 +0x1c7c
tailscale.com/wgengine.(*watchdogEngine).Reconfig.func1()
	tailscale.com@v1.98.1/wgengine/watchdog.go:176 +0x40
tailscale.com/wgengine.(*watchdogEngine).watchdogErr.func2()
	tailscale.com@v1.98.1/wgengine/watchdog.go:121 +0x28
created by tailscale.com/wgengine.(*watchdogEngine).watchdogErr in goroutine 151
	tailscale.com@v1.98.1/wgengine/watchdog.go:120 +0x1bc
```
(IP addresses and hostnames redacted.)

### Steps to reproduce

Copy new binary into place and launch:
`/userdata/tailscale/tailscaled -port 41641`

Without changing anything else, bring back a [version 1.96.4 binary](https://pkgs.tailscale.com/stable/tailscale_1.96.4_arm.tgz), and things work as usual.

### Are there any recent changes that introduced the issue?

_No response_

### OS

Linux

### OS version

Linux host 5.10.160 #1 Tue Apr 28 09:25:34 CEST 2026 armv7l GNU/Linux

### Tailscale version

1.98.1

### Other software

JetKVM system [version 0.2.8](https://github.com/jetkvm/rv1106-system/releases/tag/release%2Fv0.2.8), app [version 0.5.8](https://github.com/jetkvm/kvm/releases/tag/release%2F0.5.8).

### Bug report

_No response_

--- a/wgengine/router/osrouter/router_linux.go
+++ b/wgengine/router/osrouter/router_linux.go
@@ -490,7 +490,9 @@ func (r *linuxRouter) Set(cfg *router.Config) error {
 	// Connmark rules for rp_filter compatibility.
 	// Always enabled when netfilter is ON to handle all rp_filter=1 scenarios
 	// (normal operation, exit nodes, subnet routers, and clients using exit nodes).
-	netfilterOn := cfg.NetfilterMode == netfilterOn
+	// Gate on r.netfilterMode (actual state) rather than cfg.NetfilterMode
+	// (desired state) so we don't call into the runner when chain setup failed.
+	netfilterOn := r.netfilterMode == netfilterOn
 	switch {
 	case netfilterOn == r.connmarkEnabled:
 		// state already correct, nothing to do.
@@ -530,8 +532,8 @@ func (r *linuxRouter) Set(cfg *router.Config) error {
 		r.enableIPForwarding()
 	}
 
-	// Remove the rule to drop off-tailnet CGNAT traffic, if asked.
-	if netfilterOn || cfg.NetfilterMode == netfilterNoDivert {
+	// Remove the rule to drop off-tailnet CGNAT traffic, if needed.
+	if netfilterOn || r.netfilterMode == netfilterNoDivert {
 		var cgnatMode linuxfw.CGNATMode
 		if cfg.RemoveCGNATDropRule {
 			cgnatMode = linuxfw.CGNATModeReturn

--- a/wgengine/router/osrouter/router_linux_test.go
+++ b/wgengine/router/osrouter/router_linux_test.go
@@ -562,6 +562,10 @@ type fakeIPTablesRunner struct {
 	ipt4 map[string][]string
 	ipt6 map[string][]string
 	// we always assume ipv6 and ipv6 nat are enabled when testing
+
+	addChainsErr          error // if non-nil, AddChains returns it instead of setting up chains
+	addConnmarkSaveCalls  int
+	addExternalCGNATCalls int
 }
 
 func newIPTablesRunner(t *testing.T) linuxfw.NetfilterRunner {
@@ -794,6 +798,9 @@ func (n *fakeIPTablesRunner) DelHooks(logf logger.Logf) error {
 }
 
 func (n *fakeIPTablesRunner) AddChains() error {
+	if n.addChainsErr != nil {
+		return n.addChainsErr
+	}
 	for _, ipt := range []map[string][]string{n.ipt4, n.ipt6} {
 		for _, chain := range []string{"filter/ts-input", "filter/ts-forward", "nat/ts-postrouting"} {
 			ipt[chain] = nil
@@ -922,6 +929,7 @@ func (n *fakeIPTablesRunner) DelMagicsockPortRule(port uint16, network string) e
 }
 
 func (n *fakeIPTablesRunner) AddConnmarkSaveRule() error {
+	n.addConnmarkSaveCalls++
 	// PREROUTING rule: restore mark from conntrack
 	prerouteRule := "-m conntrack --ctstate ESTABLISHED,RELATED -j CONNMARK --restore-mark --nfmask 0xff0000 --ctmask 0xff0000"
 	for _, ipt := range []map[string][]string{n.ipt4, n.ipt6} {
@@ -970,6 +978,7 @@ func buildExternalCGNATRules(mode linuxfw.CGNATMode, tunname string) ([]iptRule,
 }
 
 func (n *fakeIPTablesRunner) AddExternalCGNATRules(mode linuxfw.CGNATMode, tunname string) error {
+	n.addExternalCGNATCalls++
 	rules, err := buildExternalCGNATRules(mode, tunname)
 	if err != nil {
 		return err
@@ -1548,5 +1557,55 @@ func TestUpdateMagicsockPortChange(t *testing.T) {
 	if hasOldRule {
 		t.Errorf("firewall rule for OLD port 12345 still exists (should be deleted).\nFound: %s\nAll rules: %v",
 			oldPortRule, nfr.ipt4["filter/ts-input"])
+	}
+}
+
+// TestSetSkipsNetfilterAddonsWhenSetupFails verifies that Set does not invoke
+// rule-management methods that depend on the ts-* chains existing when chain
+// setup failed.
+func TestSetSkipsNetfilterAddonsWhenSetupFails(t *testing.T) {
+	nfr := newIPTablesRunner(t).(*fakeIPTablesRunner)
+	nfr.addChainsErr = errors.New("kernel lacks netfilter support")
+
+	bus := eventbus.New()
+	defer bus.Close()
+	mon, err := netmon.New(bus, logger.Discard)
+	if err != nil {
+		t.Fatal(err)
+	}
+	mon.Start()
+	defer mon.Close()
+
+	fake := NewFakeOS(t)
+	ht := health.NewTracker(bus)
+	r, err := newUserspaceRouterAdvanced(logger.Discard, "tailscale0", mon, fake, ht, bus)
+	if err != nil {
+		t.Fatalf("newUserspaceRouterAdvanced: %v", err)
+	}
+	lr := r.(*linuxRouter)
+	lr.nfr = nfr
+	if err := lr.Up(); err != nil {
+		t.Fatalf("Up: %v", err)
+	}
+	defer lr.Close()
+
+	cfg := &Config{
+		LocalAddrs:    mustCIDRs("100.101.102.103/10"),
+		NetfilterMode: netfilterOn,
+	}
+	// Set must return an error (chain setup failed) but must not panic.
+	if err := lr.Set(cfg); err == nil {
+		t.Fatal("Set returned nil; want error because AddChains failed")
+	}
+	if lr.netfilterMode != netfilterOff {
+		t.Errorf("netfilterMode = %v; want netfilterOff after failed AddChains", lr.netfilterMode)
+	}
+	if nfr.addConnmarkSaveCalls != 0 {
+		t.Errorf("AddConnmarkSaveRule called %d times; want 0 when chain setup failed",
+			nfr.addConnmarkSaveCalls)
+	}
+	if nfr.addExternalCGNATCalls != 0 {
+		t.Errorf("AddExternalCGNATRules called %d times; want 0 when chain setup failed",
+			nfr.addExternalCGNATCalls)
 	}
 }

